src/traversability_boundary.cpp

```diff
@@ -0,0 +1,592 @@
+  return state == TraversabilityState::INTRAVERSABLE ||
+         (optimistic && state == TraversabilityState::UNKNOWN);
+}
+
+bool isIntraversable(const TraversabilityStates& states, bool optimistic) {
+  for (const auto& state : states) {
+    if (isIntraversable(state, optimistic)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void fuseStates(const TraversabilityState from,
+                TraversabilityState& to,
+                bool pessimistic) {
+  if (from == TraversabilityState::TRAVERSED || to == TraversabilityState::TRAVERSED) {
+    to = TraversabilityState::TRAVERSED;
+    return;
+  }
+  if (from == TraversabilityState::INTRAVERSABLE ||
+      to == TraversabilityState::INTRAVERSABLE) {
+    to = TraversabilityState::INTRAVERSABLE;
+    return;
+  }
+  if (pessimistic) {
+    if (from == TraversabilityState::UNKNOWN) {
+      to = TraversabilityState::UNKNOWN;
+    }
+  } else if (from == TraversabilityState::TRAVERSABLE) {
+    to = TraversabilityState::TRAVERSABLE;
+  }
+}
+
+void fuseStates(const TraversabilityStates& from,
+                TraversabilityStates& to,
+                bool pessimistic) {
+  if (to.size() < from.size()) {
+    to.resize(from.size());
+  }
+  for (size_t i = 0; i < from.size(); ++i) {
+    fuseStates(from[i], to[i], pessimistic);
+  }
+}
+
+void fuseStates(const TraversabilityState from,
+                TraversabilityStates& to,
+                bool pessimistic) {
+  for (auto& state : to) {
+    fuseStates(from, state, pessimistic);
+  }
+}
+
+std::pair<size_t, size_t> computeMinMaxTraversability(
+    const TraversabilityStates& states) {
+  // Find the minimum and maximum traversable width.
+  size_t min_traversable = 0;
+  size_t max_traversable = 0;
+  size_t current_traversable = 0;
+  size_t current_unknown = 0;
+  for (const auto& state : states) {
+    if (state == TraversabilityState::INTRAVERSABLE) {
+      current_traversable = 0;
+      current_unknown = 0;
+    } else if (state == TraversabilityState::UNKNOWN) {
+      current_unknown++;
+      current_traversable = 0;
+    } else {
+      current_traversable++;
+      current_unknown++;
+    }
+    min_traversable = std::max(min_traversable, current_traversable);
+    max_traversable = std::max(max_traversable, current_unknown);
+  }
+  return {min_traversable, max_traversable};
+}
+
+bool simplifyTraversabilityStates(TraversabilityStates& states) {
+  if (states.size() <= 1) {
+    return false;
+  }
+
+  for (size_t i = 1; i < states.size(); ++i) {
+    if (states[i] != states[0]) {
+      return false;
+    }
+  }
+  states.resize(1);
+  return true;
+}
+
+bool filterTraversabilityStates(TraversabilityStates& states, size_t width) {
+  if (width <= 1) {
+    return false;
+  }
+
+  if (states.size() < width) {
+    states = TraversabilityStates(width, TraversabilityState::INTRAVERSABLE);
+    return true;
+  }
+
+  // TODO(lschmid): Implement.
+  return false;
```

Comment:
Preference for no dead code if possible (I'd comment out the rest of the function if it's temporarily not enabled)